compiler/p/codegen/OMRInstOpCode.hpp

```diff
@@ -164,33 +169,85 @@ class InstOpCode: public OMR::InstOpCode
    PPCInstructionFormat getFormat() { return getMetaData().format; }
    const char* getMnemonicName() { return getMetaData().name; }
 
-   int8_t getBinaryLength()
+   PPCInstructionFormat getTemplateFormat()
       {
       switch (getFormat())
          {
          case FORMAT_NONE:
-            return 0;
+            return FORMAT_NONE;
+         case FORMAT_DIRECT_PREFIXED:
+         case FORMAT_RT_D34_RA_R:
+         case FORMAT_FRT_D34_RA_R:
+         case FORMAT_VRT_D34_RA_R:
+         case FORMAT_XT5_D34_RA_R:
+         case FORMAT_RS_D34_RA_R:
+         case FORMAT_FRS_D34_RA_R:
+         case FORMAT_VRS_D34_RA_R:
+         case FORMAT_XS5_D34_RA_R:
+            return FORMAT_DIRECT_PREFIXED;
          default:
+            return FORMAT_DIRECT;
+         }
+      }
+
+   int8_t getBinaryLength()
+      {
+      switch (getTemplateFormat())
+         {
+         case FORMAT_NONE:
+            return 0;
+         case FORMAT_DIRECT:
             return 4;
+         case FORMAT_DIRECT_PREFIXED:
+            return 8;
+         default:
+            TR_ASSERT_FATAL(false, "Invalid template format for %s", getMnemonicName());
          }
       }
 
    int8_t getMaxBinaryLength()
       {
-      return getBinaryLength();
+      switch (getTemplateFormat())
+         {
+         case FORMAT_NONE:
+            return 0;
+         case FORMAT_DIRECT:
+            return 4;
+         case FORMAT_DIRECT_PREFIXED:
+            // Prefixed instructions can't cross a 64-byte boundary, so we may need to emit a nop
+            // to avoid this.
+            return 12;
+         default:
+            TR_ASSERT_FATAL(false, "Invalid template format for %s", getMnemonicName());
+         }
       }
 
-   uint8_t *copyBinaryToBuffer(uint8_t *cursor)
+   uint8_t *copyBinaryToBuffer(uint8_t *byteCursor)
       {
-      switch (getFormat())
+      uint32_t *cursor = reinterpret_cast<uint32_t*>(byteCursor);
+      switch (getTemplateFormat())
          {
          case FORMAT_NONE:
             break;
-         default:
-            *reinterpret_cast<uint32_t*>(cursor) = metadata[_mnemonic].opcode;
+         case FORMAT_DIRECT:
+            *cursor = metadata[_mnemonic].opcode;
+            break;
+         case FORMAT_DIRECT_PREFIXED:
+            // Prefixed instructions can't cross a 64-byte boundary, so we may need to emit a nop
+            // to avoid this.
+            if (reinterpret_cast<uintptr_t>(cursor + 1) % 64 == 0)
+               {
+               *cursor = metadata[OMR::InstOpCode::nop].opcode;
+               cursor++;
+               byteCursor += 4;
+               }
```

Comment:
I agree that this is not ideal, but unfortunately doing this in a nice manner would require generating/encoding instructions ahead of the current instruction from within this function. This is a problem for two reasons:

1. This function's parameters do not give sufficient information to allow us to do this and it cannot easily be changed to have this information, as not all uses of this function have an associated `TR::Instruction` object at all.
2. Generating instructions _before_ the current instruction means that we can't rely on the binary encoder loop to handle things automatically (as we do with alignment nop instructions), so we need to manually initiate binary encoding of the generated instruction.

Certainly, neither of these are impossible obstacles, but they're not completely trivial either. I'd certainly like for this to be fixed at some point, but I think it's somewhat beyond the scope of this PR to do so. I've opened #5333 to track this so we don't forget to improve this at some point.

As for scripts (specifically Gita's tick attribution script), we shouldn't be relying on all ticks showing up exactly at the address the instruction was encoded at in the first place. Problems of this type can be easily avoided by checking `tickAddr >= instructionAddr && tickAddr < instructionAddr + instructionLength` instead of naively assuming that all ticks will show up exactly at the one address. In fact, we already had logic that would have broken such assumptions before, e.g. pre-refactor MemoryReference expansion.

Reply:
Existing undesirability is no reason to aggravate it.  It is fine to me with a new item to improve this in the future.

> we shouldn't be relying on all ticks showing up exactly at the address the instruction was encoded at in the first place.

That is not a good reason.  Be-able-to-match-back is our essential goal, notwithstanding the fact that the script can be improved (at least don't attribute it to unknown block or something).

Reply:
FYI: the script does handle missed instructions.

Reply:
Apart from how this affects some script, could you please provide some background on P10 prefixed instructions and the key design points of their implementation in OMR? I think it can be in the PR description or in a separate issue.

Reply:
@gita-omr Hmm, I did put a decent amount of information about why these changes were made and the way this works in the commit messages, but this information could easily get lost over time and may not cover everything you're looking for. Perhaps this is a good time to consider adding some documentation to the `doc` directory for this? I've opened #5340 to cover documenting the design surrounding both the way prefixed instructions are handled and the new binary encoder in general, since that's something I'd like to do anyway.

Does the info in the commit messages cover what you're looking for or is there something that you think needs expanding upon?

Reply:
@aviansie-ben thanks for opening the issue. Both, new binary encoder and P10 support are great pieces of work and deserve some design docs and documentation. I would suggest to separate the two, and in the design for P10 support describe (at least) the following:

1. prefixed instructions introduced  in P10
2. 4 types of prefixes and their meaning
3. how prefixed opcodes will be presented in the opcodes table, e.g. as  completely separate opcodes, separate but referring to their non-prefixed counterparts etc. (note the comment from @zl-wang about prefixes 2 and 3 using existing suffixes)
4. how prefixed instructions will be represented in the TR_Instructions hierarchy
5. how memory references with 34-bit displacements are going to be represented, created, expanded, etc.

Reply:
For the most part, I don't have anything of particular interest to say on many of these topics. The entire design here is done in such a way as to make prefixed instructions work as closely to non-prefixed instructions as possible outside the binary encoder, i.e. there are no special `TR::Instruction` subclasses for them and displacements are represented within the compiler in the same manner as we did previously, so I can only really talk about the lack of change. The only particular change here that's really visible outside of the binary encoder is that the offset of a `MemoryReference` is now an `int64_t` to accommodate larger offsets. The continuing existence of the `MemoryReference` constructor taking an `int32_t` is a wart I intend to eliminate rather quickly, as we already have an item for using factory methods for memory references (#5172), and any lingering quirks due to the switch should go away after making this change, so I wouldn't consider this to be part of the design of the POWER10 support.

As for 1, I don't really know how in-depth you'd want me to go. As far as I'm concerned, the only things I can really comment on here are the way prefixed instructions are encoded by the binary encoder (and some basic background, e.g. cannot cross 64-byte boundaries, that is only relevant to the binary encoder) and how PC-relative memory references will be represented. The PR for adding PC-relative memory references is already done but cannot be submitted as it is dependent on this change getting in first. The way these are represented is relatively straightforward and can be easily explained at that time.

In terms of 2 and 3, I still hold that the types of prefixes are irrelevant as far as the compiler is concerned. The only part that could possibly care about this is the binary encoder. However, the binary encoder works on the basic principle of copying a "template instruction" that we directly provide and then filling in fields within that instruction, meaning that it really doesn't care how the template instruction is derived (we just copy it from the ISA document without needing to understand the details). I also still do not intend on implementing Julian's suggestion of copying the suffix word from another entry in the properties table, as I believe this would give minimal benefit and would be needlessly confusing for anybody not well-versed in the hardware for the reasons I already outlined in https://github.com/eclipse/omr/pull/5317#discussion_r444948742. Instead, I intend to continue to treat them as if they're completely separate opcodes, so there's not much to discuss there in terms of design.